### PR TITLE
Nonlinear sum refomulation of objective of MINLP 

### PR DESCRIPTION
**Pull request by ZedongPeng**
## Summary/Motivation:
This PR adds a new configuration named `partition_obj_nonlinear_terms` in MindtPy.
Previously, if MindtPy detects a nonlinear objective function in the MINLP, the nonlinear objective will be rewritten as a single constraint using epigraph reformulation. 
In this PR, if `partition_obj_nonlinear_terms=True`, the nonlinear objective will be first partitioned into a **linear expr** and a **nonlinear expr**. 

- The linear expr will be reformulated as one epigraph constraint. 
- If the nonlinear expr is an `EXPR.SumExpression`, the nonlinear expr will be further partitioned into **subexprs**. **Each subexpr will be reformulated as one epigraph constraint.**
- If the nonlinear expr is nonseparable, the nonlinear expr will be directly reformulated as one epigraph constraint.

I also test the efficiency improvement brought by `partition_obj_nonlinear_terms` in [netmod_dol2](https://www.minlplib.org/netmod_dol2.html). 
**netmod_dol2** is a MBQP with a separable quadratic objective and all linear constraints. 

- Result when `partition_obj_nonlinear_terms=False`
```
=============================================================================================
 Iteration | Subproblem Type | Objective Value | Lower Bound | Upper Bound |   Gap   | Time(s)

         -       Relaxed NLP         -0.651008     -0.651008           inf      nan%      3.73
         1              MILP         -0.641166     -0.641166           inf      nan%      4.67
*        1         Fixed NLP         -0.252825     -0.641166     -0.252825   153.60%      6.86
         2              MILP         -0.639299     -0.639299     -0.252825   152.86%      8.10
         3              MILP          -0.63662      -0.63662     -0.252825   151.80%     16.61
         3         Fixed NLP       2.66651e-06      -0.63662     -0.252825   151.80%     18.12
         4              MILP          -0.63307      -0.63307     -0.252825   150.40%     19.77
         5              MILP         -0.632281     -0.632281     -0.252825   150.09%     24.97
*        5         Fixed NLP         -0.455101     -0.632281     -0.455101    38.93%     27.27
         6              MILP         -0.629159     -0.629159     -0.455101    38.25%     28.63
*        6         Fixed NLP         -0.470152     -0.629159     -0.470152    33.82%     30.95
         7              MILP         -0.618455     -0.618455     -0.470152    31.54%     32.52
*        7         Fixed NLP         -0.510599     -0.618455     -0.510599    21.12%     34.63
         8              MILP         -0.609398     -0.609398     -0.510599    19.35%     36.76
         8         Fixed NLP         -0.501943     -0.609398     -0.510599    19.35%     38.26
         9              MILP         -0.608737     -0.608737     -0.510599    19.22%     41.12
         9         Fixed NLP         -0.472431     -0.608737     -0.510599    19.22%     42.59
        10              MILP         -0.600096     -0.600096     -0.510599    17.53%     44.40
*       10         Fixed NLP         -0.526758     -0.600096     -0.526758    13.92%     47.07
        11              MILP         -0.598929     -0.598929     -0.526758    13.70%     49.26
        11         Fixed NLP         -0.494913     -0.598929     -0.526758    13.70%     50.38
        12              MILP         -0.594527     -0.594527     -0.526758    12.87%     52.81
*       12         Fixed NLP         -0.530215     -0.594527     -0.530215    12.13%     55.24
        13              MILP         -0.592132     -0.592132     -0.530215    11.68%     57.34
*       13         Fixed NLP         -0.533671     -0.592132     -0.533671    10.95%     59.53
        14              MILP         -0.587134     -0.587134     -0.533671    10.02%     61.96
        14         Fixed NLP         -0.500857     -0.587134     -0.533671    10.02%     63.51
        15              MILP         -0.584227     -0.584227     -0.533671     9.47%     66.71
        15         Fixed NLP         -0.361079     -0.584227     -0.533671     9.47%     68.22
        16              MILP          -0.58283      -0.58283     -0.533671     9.21%     70.56
        16         Fixed NLP         -0.512815      -0.58283     -0.533671     9.21%     71.82
        17              MILP         -0.580309     -0.580309     -0.533671     8.74%     76.60
        17         Fixed NLP         -0.529742     -0.580309     -0.533671     8.74%     78.18
        18              MILP         -0.579794     -0.579794     -0.533671     8.64%     81.45
*       18         Fixed NLP         -0.554317     -0.579794     -0.554317     4.60%     83.61
        19              MILP         -0.579691     -0.579691     -0.554317     4.58%     86.16
        19         Fixed NLP         -0.503617     -0.579691     -0.554317     4.58%     87.79
        20              MILP         -0.574878     -0.574878     -0.554317     3.71%     91.09
        20         Fixed NLP         -0.533911     -0.574878     -0.554317     3.71%     92.55
        21              MILP         -0.574846     -0.574846     -0.554317     3.70%     94.38
        21         Fixed NLP         -0.546846     -0.574846     -0.554317     3.70%     96.07
        22              MILP          -0.57285      -0.57285     -0.554317     3.34%    100.18
        22         Fixed NLP         -0.521736      -0.57285     -0.554317     3.34%    101.74
        23              MILP         -0.571075     -0.571075     -0.554317     3.02%    107.33
        23         Fixed NLP         -0.548535     -0.571075     -0.554317     3.02%    109.39
        24              MILP          -0.56958      -0.56958     -0.554317     2.75%    113.57
        24         Fixed NLP         -0.533105      -0.56958     -0.554317     2.75%    115.22
        25              MILP         -0.569329     -0.569329     -0.554317     2.71%    119.44
        25         Fixed NLP          -0.52699     -0.569329     -0.554317     2.71%    120.74
        26              MILP         -0.568608     -0.568608     -0.554317     2.58%    125.33
        26         Fixed NLP         -0.547373     -0.568608     -0.554317     2.58%    126.63
        27              MILP         -0.565317     -0.565317     -0.554317     1.98%    130.25
*       27         Fixed NLP         -0.559711     -0.565317     -0.559711     1.00%    132.85
        28              MILP         -0.565123     -0.565123     -0.559711     0.97%    136.90
        28         Fixed NLP         -0.513312     -0.565123     -0.559711     0.97%    138.47
        29              MILP         -0.563682     -0.563682     -0.559711     0.71%    143.10
        29         Fixed NLP         -0.495966     -0.563682     -0.559711     0.71%    144.50
        30              MILP         -0.563294     -0.563294     -0.559711     0.64%    148.53
        30         Fixed NLP         -0.547946     -0.563294     -0.559711     0.64%    150.35
        31              MILP         -0.562496     -0.562496     -0.559711     0.50%    155.00
        31         Fixed NLP         -0.534802     -0.562496     -0.559711     0.50%    156.25
        32              MILP          -0.56114      -0.56114     -0.559711     0.26%    161.26
        32         Fixed NLP          -0.55782      -0.56114     -0.559711     0.26%    163.19
        33              MILP         -0.560039     -0.560039     -0.559711     0.06%    167.59
MindtPy exiting on bound convergence. (UB: -0.5597114424292737 - LB: -0.5600393710944893)/ (1e-10+|bestinteger|:0.5597114424292737) <= relative tolerance: 0.001
 =============================================================================================
```

- Result when `partition_obj_nonlinear_terms=True`
```
=============================================================================================
 Iteration | Subproblem Type | Objective Value | Lower Bound | Upper Bound |   Gap   | Time(s)

         -       Relaxed NLP         -0.651008     -0.651008           inf      nan%      3.22
         1              MILP         -0.604047     -0.604047           inf      nan%      4.97
*        1         Fixed NLP         -0.527651     -0.604047     -0.527651    14.48%      7.23
         2              MILP         -0.590167     -0.590167     -0.527651    11.85%      9.44
         2         Fixed NLP         -0.416213     -0.590167     -0.527651    11.85%     10.85
         3              MILP         -0.582795     -0.582795     -0.527651    10.45%     12.97
         3         Fixed NLP         -0.408447     -0.582795     -0.527651    10.45%     14.41
         4              MILP         -0.576164     -0.576164     -0.527651     9.19%     17.09
*        4         Fixed NLP         -0.546708     -0.576164     -0.546708     5.39%     19.37
         5              MILP          -0.56642      -0.56642     -0.546708     3.61%     21.87
*        5         Fixed NLP         -0.560007      -0.56642     -0.560007     1.15%     24.23
         6              MILP         -0.565906     -0.565906     -0.560007     1.05%     26.87
         6         Fixed NLP         -0.353267     -0.565906     -0.560007     1.05%     28.40
         7              MILP         -0.560431     -0.560431     -0.560007     0.08%     30.75
MindtPy exiting on bound convergence. (UB: -0.5600074467539216 - LB: -0.5604307498614925)/ (1e-10+|bestinteger|:0.5600074467539216) <= relative tolerance: 0.001
 =============================================================================================
```

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/gdpopt/util.py
Comment (bernalde) [inline pyomo/contrib/gdpopt/util.py:137]: This is an open question, but should we have MCPP as the default and FBBT as the fallback? I get that both are contrib packages but recent PRs make the FBBT installation "easier" than MC++, and the previous version was completely included in Pyomo. Any thoughts @michaelbynum ?

File: pyomo/contrib/mindtpy/MindtPy.py
Comment (bernalde) [inline pyomo/contrib/mindtpy/MindtPy.py:120]: Should this TODO be left behind on this PR or solved here?

File: pyomo/contrib/mindtpy/config_options.py
Comment (bernalde) [inline pyomo/contrib/mindtpy/config_options.py:207]: Having this option on by default is supported by benchmarks? Is this the case with SHOT?

File: pyomo/contrib/mindtpy/feasibility_pump.py
Comment (bernalde) [inline pyomo/contrib/mindtpy/feasibility_pump.py:80]: Add doi of this paper here https://doi.org/10.1007/s10107-012-0608-x

**Review by bernalde — CHANGES_REQUESTED** [review]
Zedong, great work as usual in cleaning up MindtPy's code and adding new functionality.
My main concerns are: 
1. You are making the disaggregation the new default. Does it actually work better?
2. This is using by default MC++ to obtain bounds on the objective. Should it not be FBBT?
3. There are no new tests for this new functionality. Adding a test to see if this works as intended is encouraged.

File: pyomo/contrib/gdpopt/util.py
Comment (michaelbynum) [reply pyomo/contrib/gdpopt/util.py:137]: @bernalde In this context, the pure python version should work fine, which would not require any compiled extensions. If you already use MC++ or FBBT elsewhere, then that may influence your decision.

Comment (michaelbynum) [reply pyomo/contrib/gdpopt/util.py:137]: By the way, the pure python implementation of FBBT will not go away (exactly for this reason - it can be cumbersome for some users to compile extensions). I expect the Appsi implementation to be a fast alternative to the pure python version. Also, in my experience, the compiled version is only faster if you need to use it iteratively.

File: pyomo/contrib/mindtpy/config_options.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/config_options.py:207]: We can add this option in the benchmark code. By default, SHOT will automatically partition obj nonlinear terms if the nonlinear terms are convex.

Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/config_options.py:207]: We can add this option in the benchmark code. SHOT will automatically partition obj nonlinear terms if the nonlinear terms are convex by default.

Comment (bernalde) [reply pyomo/contrib/mindtpy/config_options.py:207]: It works well for convex cases, but since we do not check for convexity, I do not know whether this partition would work well with nonconvex and augmented penalty for example.
Let's benchmark it and decide based on it.

File: pyomo/contrib/mindtpy/MindtPy.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/MindtPy.py:120]: This TODO will be solved here.

File: pyomo/contrib/gdpopt/util.py
Comment (bernalde) [reply pyomo/contrib/gdpopt/util.py:137]: Well, this solves it. Let's use FBBT as default and MC++ as a fallback option.

File: pyomo/contrib/mindtpy/MindtPy.py
Comment (bernalde) [reply pyomo/contrib/mindtpy/MindtPy.py:120]: Alright, so let's wait for it to be solved before merging. I don't know why should it fail there though... I'll think about it and see if I can find a solution

File: pyomo/contrib/gdpopt/util.py
Comment (ZedongPeng) [reply pyomo/contrib/gdpopt/util.py:137]: No problem.

File: pyomo/contrib/mindtpy/config_options.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/config_options.py:207]: No problem.

File: pyomo/contrib/mindtpy/MindtPy.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/MindtPy.py:120]: It might not be a bug. I just forget the implementation details and I'm trying to recall them.

File: pyomo/contrib/gdpopt/util.py
Comment (ZedongPeng) [reply pyomo/contrib/gdpopt/util.py:137]: In MindtPy configurations, `use_mcpp = False` by default. Therefore, MindtPy already uses FBBT as default and MC++ as a fallback option. I have changed the default value of `use_mcpp` in `process_objective` function.

File: pyomo/contrib/mindtpy/config_options.py
Comment (jsiirola) [reply pyomo/contrib/mindtpy/config_options.py:207]: There is also the option to use SUSPECT (if present) to do convexity detection.

Comment (michaelbynum) [reply pyomo/contrib/mindtpy/config_options.py:207]: Good point @jsiirola. Suspect can do FBBT as well.

**Comment by jsiirola** [comment]
I think this PR looks good.  Once @bernalde's comment about the TODO is resolved, I think we can merge.

File: pyomo/contrib/mindtpy/MindtPy.py
Comment (bernalde) [reply pyomo/contrib/mindtpy/MindtPy.py]: This is the only part holding this PR, there is some grammar mistake in the comment `Since the distance calculated does not include these variables, ...` but more importantly, is it not yet clear if the problem has been solved or not.

Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/MindtPy.py]: This problem has been resolved. The original implementation is correct and I added more descriptions for it.

File: pyomo/contrib/mindtpy/MindtPy.py
Comment (bernalde) [inline pyomo/contrib/mindtpy/MindtPy.py:123]: Typo, `original`

File: pyomo/contrib/mindtpy/MindtPy.py
Comment (bernalde) [inline pyomo/contrib/mindtpy/MindtPy.py:124]: Typo `we`

**Review by bernalde — APPROVED** [review]
A couple of typos, but they should not stop this from being merged.

**Comment by ZedongPeng** [comment]
I already have the benchmark results. Please give me one more day to analyze the results and determine the default value for `partition_obj_nonlinear_terms` option before merging.

File: pyomo/contrib/mindtpy/tests/test_mindtpy.py
Comment (jsiirola) [inline pyomo/contrib/mindtpy/tests/test_mindtpy.py:215]: Should this message be sent out through a logger?

**Review by jsiirola — APPROVED** [review]
This should be good to go.

File: pyomo/contrib/mindtpy/tests/test_mindtpy.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/tests/test_mindtpy.py]: I just removed it.

**Comment by ZedongPeng** [comment]
I think this PR can be merged.

**Review by jsiirola — APPROVED** [review]

**Review by bernalde — APPROVED** [review]

**Review by michaelbynum — APPROVED** [review]

**Pull request merged**
